Docs: add the "How to add support for custom data types" section (#4049)
aiidateam/aiida-core: docs/source/howto/data.rst

            :return: the content of the file as a string
            """
            filename = self.get_attribute('filename')
            return self.get_object_content(filename)

To create a new instance of this data type and get its content:

.. code:: python

    node = TextFileData(filepath='/some/absolute/path/to/file.txt')
    node.get_content()  # This will return the content of the file

This example is a simplified version of the :class:`~aiida.orm.nodes.data.singlefile.SinglefileData` data class that ships with ``aiida-core``.
If this happens to be your use case (or very close to it), it is of course better to use that class, or you can sub class it and adapt it where needed.

The two new data types we have just implemented are of course trivial and not very useful, but the concepts are flexible and can easily be applied to more complex use-cases.
The following section will provide useful guidelines on how to optimally design new data types.

.. _how-to:data:plugin:design-guidelines:

Database or repository?
-----------------------

When deciding where to store a property of a data type, one has to choose between the database and the file repository.

> The database will make it possible to search in the provenance graph based on criteria based on the property, e.g. all ``NewData`` nodes where the property is greater than 0.

Author: I agree that it is not the clearest. I find it a bit difficult to explain in words without becoming too technical. Let me give the technical version and then we can see together if we can come up with a better description for the documentation.

All nodes are stored in the database in the node table. The node table has multiple columns that store the properties of each node. Most of these are properties that exist for all node types, e.g., pk, UUID, label, ctime, mtime and description. But we also need a way to store data type specific data, which is what the `attributes` column is for, which can contain any object in JSON format, so it can store any key-value pair as long as the value is serializable. So the data type specific content is stored in the "attributes" (which become immutable once the node is stored.). You can see that in AiIDA's context, a node's attributes is quite specific, as it is literally the content of the attribute column. That is why I tried to use "properties" (instead of attributes) as a more generic term to describe the "things that define the data node". 

So finally, what I mean with property is "property" of something in the generic sense, not in the Python sense, but in the way that the UUID of a node is a property. What I mean with "searching based on criteria based on the property" is simply that clearly, when storing a property of a data node in the database (as opposed to the repository) one can query for it. Not sure how clear it is to the average user that storing the data in the database enables querying. They might not have encountered the query builder at this point in time.

> The database will make it possible to search in the provenance graph based on criteria based on the property, e.g. all ``NewData`` nodes where the property is greater than 0.

Reviewer: How about this? 

> All node properties, including attributes that are stored in the database, are directly searchable as part of a database query whereas data stored in the file repository provides no introspection. What this means is that, for example, it is possible to search for all nodes where a particular database-stored integer attribute falls into a certain value range, but the same value stored in a file within the file repository would not be directly searchable in this way.

> The database will make it possible to search in the provenance graph based on criteria based on the property, e.g. all ``NewData`` nodes where the property is greater than 0.

Author: With a few modifications I think that could work:
> All node properties that are stored in the database (such as the attributes), are directly searchable as part of a database query, whereas data stored in the file repository cannot be queried for.
What this means is that, for example, it is possible to search for all nodes where a particular database-stored integer attribute falls into a certain value range, but the same value stored in a file within the file repository would not be directly searchable in this way.

> The downside is that storing too much information in the database can make it sluggish.

Reviewer: ```suggestion
However storing large amounts of binary data within the database comes at the cost of slowing down database queries.
```

Therefore, big data (think large files), whose content does not necessarily need to be queried for, is better stored in the file repository.

> Of course a data type may need to store multiple properties of varying character, but both storage locations can safely be used in parellel.

Reviewer: ```suggestion
A data type may safely use both the database and file repository in parallel for individual properties.
```

> When choosing the database as the storage location, the properties should be stored using the node *attributes*.

Reviewer: ```suggestion
Properties stored as part of the node's *attributes* are stored in the database.
```
Is that right? I'm trying to simplify the sentence, but I'm not 100% sure whether I got cause and effect right.

Author: Technically it is correct yes. What I think is missing here is a direct reference to the actual methods of the `Node` class. Without those, the term `attributes` is too vague I think. Maybe I can add a sentence like:
```
The node class has various methods to set its attributes, such as :py:`~aiida.orm.node.Node.set_attribute` and :py:`~aiida.orm.node.Node.set_attribute_many`
```

Reviewer: Maybe you can provide an example that references those methods?
